memory.go

```diff
@@ -0,0 +1,227 @@
+
+// Memory implements accessing a Map's memory without making any syscalls.
+type Memory struct {
+	b  []byte
+	ro bool
+}
+
+func newMemory(fd, s int, ro bool) (*Memory, error) {
+	// Allocate a page-aligned span of memory on the Go heap.
+	alloc, size, err := allocate(s)
+	if err != nil {
+		return nil, fmt.Errorf("allocating memory: %w", err)
+	}
+
+	flags := unix.PROT_READ | unix.PROT_WRITE
+	if ro {
+		flags = unix.PROT_READ
+	}
+
+	// Map the bpf map memory over the Go heap. This will result in the following
+	// mmap layout in the process' address space (0xc000000000 is a span of Go
+	// heap), visualized using pmap:
+	//
+	// Address           Kbytes     RSS   Dirty Mode  Mapping
+	// 000000c000000000    1824     864     864 rw--- [ anon ]
+	// 000000c0001c8000       4       4       4 rw-s- [ anon ]
+	// 000000c0001c9000    2268      16      16 rw--- [ anon ]
+	//
+	// This will break up the Go heap, but as long as the runtime doesn't try to
+	// move our allocation around, this is safe for as long as we hold a reference
+	// to our allocated object.
+	//
+	// Use MAP_SHARED to make sure the kernel sees any writes we do, and MAP_FIXED
+	// to ensure the mapping starts exactly at the address we requested. If alloc
+	// isn't page-aligned, the mapping operation will fail.
+	if _, err = unix.MmapPtr(fd, 0, alloc, uintptr(size),
+		flags, unix.MAP_SHARED|unix.MAP_FIXED); err != nil {
+		return nil, fmt.Errorf("setting up memory-mapped region: %w", err)
+	}
+
+	// Set a finalizer on the heap allocation to undo the mapping before the span
+	// is collected and reused by the runtime. This has a few reasons:
+	//
+	//  - Avoid leaking memory/mappings.
+	//  - Future writes to this memory should never clobber a bpf map's contents.
+	//  - Some bpf maps are mapped read-only, causing a segfault if the runtime
+	//    reallocates and zeroes the span later.
+	runtime.SetFinalizer((*byte)(alloc), unmap(size))
+
+	mm := &Memory{
+		unsafe.Slice((*byte)(alloc), size),
+		ro,
+	}
+
+	return mm, nil
+}
+
+// allocate returns an unsafe.Pointer to a page-aligned section of memory on the
+// Go heap, managed by the runtime. The given size is rounded up to the nearest
+// multiple of the system's page size to ensure we're given an allocation that
+// starts on a page boundary. The size of the resulting allocation is returned.
+func allocate(size int) (unsafe.Pointer, int, error) {
+	// Memory-mapping over a piece of the Go heap is unsafe when the GC can
+	// randomly decide to move objects around, in which case the mapped region
+	// will not move along with it.
+	if heapObjectsCanMove() {
+		return nil, 0, errors.New("this Go runtime has a moving garbage collector")
+	}
+
+	// Request at least a full page from the runtime, otherwise the allocated span
+	// is likely not page-aligned, risking mapping over objects on another page.
+	// Since we use MAP_FIXED, the starting address of the mapping must be
+	// page-aligned on most architectures anyway.
```

Comment:
Pardon my ignorance, how does page-misaligned memory mapping work on those architectures?

Reply:
I'm paraphrasing `man mmap(2)` here.

```
MAP_FIXED
              Don't interpret addr as a hint: place the mapping at
              exactly that address.  addr must be suitably aligned: for
              most architectures a multiple of the page size is
              sufficient; however, some architectures may impose
              additional restrictions. [...]
```

I briefly came across one such additional restriction somewhere, but I forgot, sorry.

Here's another snippet: https://pubs.opengroup.org/onlinepubs/009604499/functions/mmap.html (this seems to be some kind of POSIX spec) 
```
[EINVAL]
    The addr argument (if MAP_FIXED was specified) or off is not a multiple of the page size as returned by sysconf(), or is considered invalid by the implementation.
```

tl;dr - `off` within the target fd must also be page-aligned. We're not using the offset parameter.

Reply:
```suggestion
	// Since we use MAP_FIXED, the starting address of the mapping must be
	// page-aligned anyway.
```
The original wording can be interpreted as some architectures being fine with non-page-aligned mmap starting address.

Reply:
Sure, the starting address being page-aligned for MAP_FIXED sounds universal indeed.